File: Javascript/Tree/M-236. Lowest Common Ancestor of a Binary Tree.ts
/** Definition for a binary tree node. */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function lowestCommonAncestor(
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode | null {
  if (!root || root === p || root === q) return root;
  const nodeFromL = lowestCommonAncestor(root.left, p, q);
  const nodeFromR = lowestCommonAncestor(root.right, p, q);
  return nodeFromL && nodeFromR ? root : nodeFromL || nodeFromR;
}

/** case 1 */
const case1Lvl4Nodes = [new TreeNode(7), new TreeNode(4)];
const case1Lvl3Nodes = [
  new TreeNode(6),
  new TreeNode(2, case1Lvl4Nodes[0], case1Lvl4Nodes[1]),
  new TreeNode(0),
  new TreeNode(8),
];
const case1Lvl2Nodes = [
  new TreeNode(5, case1Lvl3Nodes[0], case1Lvl3Nodes[1]),
  new TreeNode(1, case1Lvl3Nodes[2], case1Lvl3Nodes[3]),
];
const case1RootNode = new TreeNode(3, case1Lvl2Nodes[0], case1Lvl2Nodes[1]);

console.log(
  lowestCommonAncestor(case1RootNode, case1Lvl2Nodes[0], case1Lvl4Nodes[1]).val
); // should be 5

File: Python/Tree/M-236. Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    @classmethod
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode"
    ) -> "TreeNode":
        if root is None or root == p or root == q:
            return root
        node_from_left = Solution.lowestCommonAncestor(root.left, p, q)
        node_from_right = Solution.lowestCommonAncestor(root.right, p, q)

        if node_from_left and node_from_right:
            return root

        return node_from_left or node_from_right
